File: fxgl-samples/src/main/java/s08ai/AttackTask.java
/*
 * FXGL - JavaFX Game Library. The MIT License (MIT).
 * Copyright (c) AlmasB (almaslvl@gmail.com).
 * See LICENSE for details.
 */

package s08ai;

import com.almasb.fxgl.ai.AIControl;
import com.almasb.fxgl.ai.SingleAction;
import com.almasb.fxgl.app.FXGL;
import com.almasb.fxgl.entity.Entity;
import javafx.geometry.Point2D;

/**
 * @author Almas Baimagambetov (almaslvl@gmail.com)
 */
public class AttackTask extends SingleAction {

    @Override
    public void onUpdate(double tpf) {
        Entity player = ((BehaviorSample) FXGL.getApp()).player;

        if (player.distance(getEntity()) < 100) {
            getEntity().getComponent(AIControl.class).setBubbleMessage("Attack");
        } else {
            getEntity().getComponent(AIControl.class).setBubbleMessage("Chase");
            double speed = tpf * 60 * 5;

            Point2D vector = player.getPosition()
                    .subtract(getEntity().getPosition())
                    .normalize()
                    .multiply(speed);

            getEntity().getPositionComponent().translate(vector);
        }
    }
}

File: fxgl-engine/src/test/kotlin/com/almasb/fxgl/ai/AttackTask.kt
/*
 * FXGL - JavaFX Game Library. The MIT License (MIT).
 * Copyright (c) AlmasB (almaslvl@gmail.com).
 * See LICENSE for details.
 */

package com.almasb.fxgl.ai

import com.almasb.fxgl.ai.btree.LeafTask
import com.almasb.fxgl.ai.btree.Task
import com.almasb.fxgl.entity.Entity

/**
 *
 *
 * @author Almas Baimagambetov (almaslvl@gmail.com)
 */
class AttackTask : LeafTask<Entity>() {

    override fun execute(): Status {
        return Status.SUCCEEDED
    }

    override fun copyTo(task: Task<Entity>): Task<Entity> {
        return task
    }
}